### PR TITLE
Code for more flexible interpolation schemes

### PR DESCRIPTION
**Pull request by duncan-brown**

### CODE REVIEW

**Comment by duncan-brown** [comment]
@cdcapano any objections to @soumide1102 merging this?

**Comment by duncan-brown** [comment]
@soumide1102 has tested this against a bank and found that the mismatches are all very small, as expected. 

match-v2romvspartialdecrom-fullbank

**Comment by soumide1102** [comment]
@duncan-brown The plot that I made for the bank is showing mismatches between the original SEOBNRv2_ROMs from the LAL code and the partially decompressed ROMs(parameter space interpolation in SEOBNRv2_ROM LAL code, frequency space interpolation using scipy linear interpolation). This PR does not yet include the modifications I made in ```compress.py``` and ```pycbc_compress_bank``` to perform the partial decompression. Performing the partial decompression also includes changes made in the LAL code. 

I had pushed the changes to include the partial decompression to your branch if you want to take a look, but they have not been pushed up to this PR currently. This PR currently includes the modification that we did to Collin's compression method and decompression using inline_linear or scipy.

**Comment by duncan-brown** [comment]
@soumide1102 I think I merged this https://github.com/duncan-brown/pycbc/pull/3

There are no currently open pull request on my repository.

File: bin/pycbc_compress_bank [hunk 1]
```diff
@@ -72,11 +72,17 @@ parser.add_argument("--tolerance", type=float, default=0.001,
...
-                     "or any interpolation recognized by scipy's interp1d "
-                     "kind argument. Default is linear.")
+                     "waveforms for checking tolerance. Options are "
+                     "'linear_same_freq_points', which assumes that the "
+                     "frequency samples of the amplitude and phase "
+                     "interpolants are the same, or any interpolation "
+                     "recognized by scipy's interp1d kind argument, which "
+                     "allows the frequency interpolant samples to be "
+                     "different for the amplitude and phase. "
+                     "Default is linear_same_freq_points.")
```
Comment (cdcapano) [inline bin/pycbc_compress_bank]: The help message doesn't match the code. The default is `linear` but the help message says `linear_same_freq_points`. In any event, this should just be `linear`. This argument's only purpose is to determine the algorithm used for interpolating points, not how those points are determined. The latter is determined by the `compression-algorithm` argument. It doesn't make sense to use different amplitude/phase points for `spa` or `mchirp` compression; likewise it doesn't make sense to use the same points for the compression you are working on. When you add your compression algorithm, you can add the switch to different points based on what `compression-algorithm` is set to.

File: pycbc/waveform/compress.py [hunk 2]
```diff
@@ -423,7 +430,7 @@ def compress_waveform(htilde, sample_points, tolerance, interpolation,
     'double': numpy.float64
 }
 
-def fd_decompress(amp, phase, sample_frequencies, out=None, df=None,
+def fd_decompress(amp, phase, amp_freq, phase_freq, out=None, df=None,
```
Comment (cdcapano) [inline pycbc/waveform/compress.py]: This isn't what we agreed to in #1083. Only the frequency points for the phase should be required. Amplitude frequency points should be a keyword argument, with the default set to None. So, have a required argument called `sample_frequencies` that is (assumed to be) both for phase and amplitude, and a keyword `amp_sample_frequencies` that, if specified, will be used for the amplitude, with the other only being used for the phase.

File: pycbc/waveform/compress.py [hunk 3]
```diff
@@ -259,7 +262,10 @@ def compress_waveform(htilde, sample_points, tolerance, interpolation,
         comp_amp = amp.take(sample_index)
         comp_phase = phase.take(sample_index)
         # update the vecdiffs and mismatch
-        hdecomp = fd_decompress(comp_amp, comp_phase, sample_points,
+        amp_sample_points = None
```
Comment (cdcapano) [inline pycbc/waveform/compress.py]: You shouldn't need to specify `amp_sample_points` here; see my comment below.

**Comment by soumide1102** [comment]
@duncan-brown I just saw that you have merged that last commit into your branch, but I don't see that commit pushed up to this PR. ```compress.py``` in this PR does not contain ```partial_compress_rom``` .

File: pycbc/waveform/compress.py [hunk 4]
```diff
@@ -578,19 +622,21 @@ class CompressedWaveform(object):
...
     Class Methods
     -------------
     from_hdf :
         Loads a compressed waveform from the given open hdf file.
     """
-    
-    def __init__(self, sample_points, amplitude, phase,
+
+    def __init__(self,
+            amplitude_freq_points, phase_freq_points, amplitude, phase,
```
Comment (cdcapano) [inline pycbc/waveform/compress.py]: As above, only a single `frequency_points` should be required, with an optional `amp_frequency_points` keyword argument set to None. Please also change the documentation string to reflect this.

**Comment by duncan-brown** [comment]
@cdcapano there is no new *compression* here, that is implemented by only partially decompressing the ROM waveforms in LAL. See https://bugs.ligo.org/redmine/issues/4825

The way that Michael implemented the ROM compression in LAL requires that the amplitude and phase interpolants be sampled at different frequency points. This patch allows that to happen.

**Comment by duncan-brown** [comment]
@soumide1102 please address Collin's comments and submit a new pull request. I think it's better if you do this directly against https://github.com/ligo-cbc/pycbc rather than going though my repository.

Given that this pull is against my repo, I am closing it and you can open a new one once you have addressed Collin's changes.

**Comment by soumide1102** [comment]
@cdcapano Yes, the PR has not been updated yet to add in the new compression algorithm that makes the plot that Duncan had attached. 
When updating the code for adding in the new compression algorithm, I am trying to separate the ```--interpolation``` options ```linear``` from ```inline_linear``` because scipy's linear interpolation requires the ```kind``` to be specified as ```linear```. If the user wants to perform ```spa_compression``` which generates one set of sample_points and then chooses to do scipy linear interpolation, the code would not be able to distinguish between inline_linear and scipy linear if both the methods are referred to as ```linear```. Thus one would only have to perform inline_linear with spa_compression in the current version of the code.

**Pull request closed without merging**